src/streaming/HTTPLoader.js

```diff
@@ -0,0 +1,313 @@
+            url: modifiedUrl,
+            method: verb,
+            withCredentials: withCredentials,
+            request: request,
+            onload: onload,
+            onend: onloadend,
+            onerror: onloadend,
+            progress: progress,
+            onabort: onabort
+        };
+
+        // Adds the ability to delay single fragment loading time to control buffer.
+        let now = new Date().getTime();
+        if (isNaN(request.delayLoadingTime) || now >= request.delayLoadingTime) {
+            // no delay - just send
+            requests.push(httpRequest);
+            httpRequest.response = loader.send(httpRequest);
+        } else {
+            // delay
+            let delayedRequest = { httpRequest: httpRequest };
+            delayedRequests.push(delayedRequest);
+            delayedRequest.delayTimeout = setTimeout(function () {
+                if (delayedRequests.indexOf(delayedRequest) === -1) {
+                    return;
+                } else {
+                    delayedRequests.splice(delayedRequests.indexOf(delayedRequest), 1);
+                }
+                try {
+                    requestStartTime = new Date();
+                    lastTraceTime = requestStartTime;
+                    requests.push(delayedRequest.httpRequest);
+                    httpRequest.response = loader.send(delayedRequest.httpRequest);
+                } catch (e) {
+                    delayedRequest.httpRequest.onerror();
+                }
+            }, (request.delayLoadingTime - now));
+        }
+    }
+
+    /**
+     * Initiates a download of the resource described by config.request
+     * @param {Object} config - contains request (FragmentRequest or derived type), and callbacks
+     * @memberof module:XHRLoader
+     * @instance
+     */
+    function load(config) {
+        if (config.request) {
+            internalLoad(
+                config,
+                mediaPlayerModel.getRetryAttemptsForType(
+                    config.request.type
+                )
+            );
+        }
+    }
+
+    /**
+     * Aborts any inflight downloads
+     * @memberof module:XHRLoader
+     * @instance
+     */
+    function abort() {
+        retryTimers.forEach(t => clearTimeout(t));
+        retryTimers = [];
+
+        delayedRequests.forEach(x => clearTimeout(x.delayTimeout));
+        delayedRequests = [];
+
+        requests.forEach(x => {
+            // abort will trigger onloadend which we don't want
+            // when deliberately aborting inflight requests -
+            // set them to undefined so they are not called
+            x.onloadend = x.onerror = x.onprogress = undefined;
+            if (x.response && x.response.abort) {
+                x.response.abort();
+            } else if (x.controller) {
+                // For firefox and edge
+                x.controller.abort();
+            } else if (x.reader) {
+                // For Chrome
+                x.reader.cancel();
```

Comment:
This can throw a TypeError which ought to be caught. For example, calling this after a network error will cause it to throw.

Reply:
Further to the above, this whole method seems to break the abstraction. Surely each request should just have an abort method or similar and the underlying functionality should be encapsulated there?